### PR TITLE
[SYCL] Add sycl::detail::bit_cast implementation

### PR DESCRIPTION
**Pull request by dm-vodopyanov**
Implementing SYCL_INTEL_bitcast extension which reinterprets the bits of an object as an object of a different data type calling the sycl::detail::bit_cast function. This function has the same specification as std::bit_cast, defined by C++20. 
sycl::detail::bit_cast calls std::bit_cast if __cpp_lib_bit_cast macro is enabled, otherwise, it calls __builtin_bit_cast builtin for oneAPI Data Parallel C++ compiler and sycl::detail::memcpy function for other compilers.

### CODE REVIEW

File: sycl/test/bit_cast/bit_cast.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t1.out
+// RUN: env SYCL_DEVICE_TYPE=HOST %t1.out
+// RUN: %CPU_RUN_PLACEHOLDER %t1.out
+// RUN: %GPU_RUN_PLACEHOLDER %t1.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
```
Comment (AlexeySachkov) [inline sycl/test/bit_cast/bit_cast.cpp:5]:
```suggestion
// RUN: %ACC_RUN_PLACEHOLDER %t1.out
```

Or, which is better, IMHO, you could rename `t1` to `t` everywhere else

File: sycl/test/bit_cast/bit_cast.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+//==---- bit_cast.cpp - SYCL bit_cast test -------------------==//
```
Comment (AlexeySachkov) [inline sycl/test/bit_cast/bit_cast.cpp:7]: There is no need in copyright headers for tests

File: sycl/test/bit_cast/bit_cast.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  return Vec[0];
+}
+
+template <typename To, typename From>
+int test(const From &ValueToConvert, const To &Expected) {
+  auto Actual = doBitCast<To>(ValueToConvert);
+  bool isActualEqualsToExpected = false;
+  if (std::is_integral<To>::value) {
+    isActualEqualsToExpected = Actual == Expected;
+  } else if ((std::is_floating_point<To>::value) || typeid(To) == typeid(cl::sycl::half)) {
```
Comment (AlexeySachkov) [inline sycl/test/bit_cast/bit_cast.cpp:48]:
```suggestion
  } else if ((std::is_floating_point<To>::value) || std::is_same<To, cl::sycl::half>::value) {
```

File: sycl/tools/CMakeLists.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD 14)
```
Comment (AlexeySachkov) [inline sycl/tools/CMakeLists.txt:1]: Looks like unrelated change, I suggest to submit it as separate PR

File: sycl/include/CL/sycl/intel/sub_group.hpp [hunk 5]
```diff
@@ -17,11 +17,11 @@
 #include <CL/sycl/detail/spirv.hpp>
 #include <CL/sycl/detail/type_traits.hpp>
 #include <CL/sycl/id.hpp>
+#include <CL/sycl/intel/bit_cast.hpp>
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/intel/sub_group.hpp:20]: Strictly speaking, this means that our implementation of sub-groups extension depends on another our extension, while there is no documented dependency in the specification.

I'm not sure how important this is and would like to hear feedback from @Pennycook here.

File: sycl/include/CL/sycl/intel/bit_cast.hpp [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+#include <CL/sycl/detail/helpers.hpp>
+
+#include <numeric>
+#include <type_traits>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+
+template <typename To, typename From>
+constexpr To bit_cast(const From &from) noexcept {
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/intel/bit_cast.hpp:21]: Interesting that extension is named `INTEL`, while the function placed into `sycl` namespace: I wonder whether we need to rename the extension spec or not.

File: sycl/include/CL/sycl/intel/sub_group.hpp (hunk 5 above)
Comment (Pennycook) [reply sycl/include/CL/sycl/intel/sub_group.hpp:20]: I agree that this could be surprising, especially if we later introduce controls to enable individual extensions.

I'm not sure this makes sense, but could we do something like:
- Move this implementation of `bit_cast` to `cl::sycl::detail`
- Expose `bit_cast` as part of `cl::sycl::intel` only when some macro like `__SYCL_INTEL_BIT_CAST__` is defined

That way our implementation can use `detail::bit_cast` wherever it wants to, including in other extensions.  But users won't see `sycl::intel::bit_cast` unless they're expecting to.

File: sycl/include/CL/sycl/intel/bit_cast.hpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  static_assert(std::is_trivially_copyable<To>::value,
+                "To must be trivially copyable");
+#if __cpp_lib_bit_cast
+  return std::bit_cast<To>(from);
+#else // __cpp_lib_bit_cast
+
+#if __has_builtin(__builtin_bit_cast)
+  return __builtin_bit_cast(To, from);
+#else  // __has_builtin(__builtin_bit_cast)
+  To to;
```
Comment (s-kanaev) [inline sycl/include/CL/sycl/intel/bit_cast.hpp:35]: `To` must be trivially-constructible in this case.

File: sycl/test/bit_cast/bit_cast.cpp (hunk 3 above)
Comment (dm-vodopyanov) [reply sycl/test/bit_cast/bit_cast.cpp]: Thanks, fixed

File: sycl/tools/CMakeLists.txt (hunk 4 above)
Comment (dm-vodopyanov) [reply sycl/tools/CMakeLists.txt]: The fix was made for use of ```std::enable_if_t``` (since C++ 14) on early steps of implementation of sycl::bit_cast. After that I replaced ```std::enable_if_t``` with ```static_assert``` as it's more user-friendly in compilation logs to find an error.
Finally, as ```static_assert``` is used, revered this.

File: sycl/test/bit_cast/bit_cast.cpp (hunk 1 above)
Comment (dm-vodopyanov) [reply sycl/test/bit_cast/bit_cast.cpp:5]: Thanks, fixed.

File: sycl/tools/CMakeLists.txt (hunk 4 above)
Comment (AlexeySachkov) [reply sycl/tools/CMakeLists.txt]: FYI, for `enable_if_t` we have `details::enable_if_t`, which works in C++11 as well

File: sycl/test/bit_cast/bit_cast.cpp (hunk 2 above)
Comment (dm-vodopyanov) [reply sycl/test/bit_cast/bit_cast.cpp]: Thanks, removed.

File: sycl/include/CL/sycl/intel/sub_group.hpp (hunk 5 above)
Comment (AlexeySachkov) [reply sycl/include/CL/sycl/intel/sub_group.hpp:20]: I agree with John, let's put actual implementation into `detail` namespace and use it from both: sub-groups and this extension

File: sycl/include/CL/sycl/intel/bit_cast.hpp (hunk 7 above)
Comment (AlexeySachkov) [reply sycl/include/CL/sycl/intel/bit_cast.hpp:35]: Good comment. However, if I understand correctly if we are using `clang` as a compiler, we should never fallback to this version.

Comment (s-kanaev) [reply sycl/include/CL/sycl/intel/bit_cast.hpp:35]: > we should never fallback to this version

We should emit an error here, then. Still, we implement some behaviour for this case. Hence, I reckon, `static_assert(std::is-trivially_constructible<To>::value, "Some good reasoning message");` is much more user-friendly than a generic compile-time error saying "your data type ain't got no viable constructor for me".

File: sycl/tools/CMakeLists.txt (hunk 4 above)
Comment (dm-vodopyanov) [reply sycl/tools/CMakeLists.txt]: Thanks, I found it then I had implemented this, but in my opinion, we should use C++14 features directly as we moved to C++14 for DPC++ runtime.

**Comment by romanovvlad** [comment]
Please, do not forget to update sycl/doc/extensions/README.md with a new status of this extension.

**Comment by romanovvlad** [comment]
Could you please add a little bit more information in the commits message: what bitcast does and what it is lower to depending on the compiler used.

File: sycl/tools/CMakeLists.txt (hunk 4 above)
Comment (AlexeySachkov) [reply sycl/tools/CMakeLists.txt]: I think that default standard which is used for compiling user apps is a bit different from standard used for our library: someone might still want to use C++11 for some reasons and it would be good to be flexible in this regard.

I'm not entirely sure that this is still actual for C++11, but for example for C++17 features I would strongly prefer that their usage is optional in our library

File: sycl/include/CL/sycl/intel/sub_group.hpp (hunk 5 above)
Comment (dm-vodopyanov) [reply sycl/include/CL/sycl/intel/sub_group.hpp:20]: > Expose bit_cast as part of ```cl::sycl::intel``` only when some macro like \_\_SYCL_INTEL_BIT_CAST\_\_ is defined

So, it should be similar to the following in SYCL_INTEL_unnamed_kernel_lambda extension? We should introduce a new compiler flag like ```-fsycl-bit_cast``` and when enabled, the extension should define the macro \_\_SYCL_INTEL_BIT_CAST\_\_?

https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/UnnamedKernelLambda/SYCL_INTEL_unnamed_kernel_lambda.asciidoc: 
> In the Intel prototype implementation of this extension at GitHub link, the extension is enabled through the compiler flag ```-fsycl-unnamed-lambda```. When enabled, the extension defines the macro \_\_SYCL_INTEL_UNNAMED_LAMBDA\_\_.

File: sycl/include/CL/sycl/intel/bit_cast.hpp (hunk 7 above)
Comment (dm-vodopyanov) [reply sycl/include/CL/sycl/intel/bit_cast.hpp]: Thanks, fixed.

File: sycl/include/CL/sycl/intel/bit_cast.hpp (hunk 6 above)
Comment (dm-vodopyanov) [reply sycl/include/CL/sycl/intel/bit_cast.hpp:19]: Done.

**Comment by dm-vodopyanov** [comment]
@romanovvlad, done.

File: sycl/include/CL/sycl/intel/sub_group.hpp (hunk 5 above)
Comment (Pennycook) [reply sycl/include/CL/sycl/intel/sub_group.hpp]: That makes sense to me.

That said...  I'd personally be okay with this PR being merged without the new compiler flag and macro, as long as the function appears in both `detail::` and `cl::sycl::intel`.  We could then revisit how to enable/disable the extension later.  What do you think @AlexeySachkov?

Comment (AlexeySachkov) [reply sycl/include/CL/sycl/intel/sub_group.hpp]: I'm okay to merge this PR without adding macro and/or compiler option for now

File: sycl/include/CL/sycl/intel/bit_cast.hpp [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+namespace intel {
+
+template <typename To, typename From>
+constexpr To bit_cast(const From &from) noexcept {
+  return sycl::detail::bit_cast<To>(from);
+}
+
+} // namespace intel
+} // namespace sycl
+} // __SYCL_INLINE_NAMESPACE(cl)
```
Comment (keryell) [inline sycl/include/CL/sycl/intel/bit_cast.hpp:25]: Missing newline

**Comment by dm-vodopyanov** [comment]
@jbrodman, @mkinsner could you please review?

File: sycl/include/CL/sycl/detail/helpers.hpp [hunk 9]
```diff
@@ -17,6 +17,7 @@
 #include <CL/sycl/detail/type_traits.hpp>
 
 #include <memory>
+#include <numeric> // std::bit_cast
```
Comment (felipe-intel) [inline sycl/include/CL/sycl/detail/helpers.hpp:20]: Isn't `std::bit_cast` defined in `<bit>`?

Comment (dm-vodopyanov) [reply sycl/include/CL/sycl/detail/helpers.hpp:20]: It is, but ```<bit>``` library is defined since C++20 and it is part of ```<numeric>``` library. DPC++ runtime builds with C++ std libraries w/o C++20 support.

Comment (felipe-intel) [reply sycl/include/CL/sycl/detail/helpers.hpp:20]: Makes sense, LGTM.

File: sycl/test/bit_cast/bit_cast.cpp [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  ReturnCode += test(static_cast<int>(2), static_cast<float>(2.8026e-45f));
+
+  std::cout << "float to int ...\n";
+  ReturnCode += test(static_cast<float>(-2.4f), static_cast<int>(-1072064102));
+
+  std::cout << "unsigned int to float ...\n";
+  ReturnCode += test(static_cast<unsigned int>(6), static_cast<float>(8.40779e-45f));
+
+  std::cout << "float to unsigned int ...\n";
+  ReturnCode += test(static_cast<float>(-2.4f), static_cast<unsigned int>(3222903194));
```
Comment (s-kanaev) [inline sycl/test/bit_cast/bit_cast.cpp:81]: There are inaccuracies in representation of floating point numbers, isn't it?
Hence, I think, it's more viable to test floating-point numbers conversion with `floating-point->integer->floating-point` and check if original and result value don't differ too much.
@dm-vodopyanov What do you think about it?

Comment (dm-vodopyanov) [reply sycl/test/bit_cast/bit_cast.cpp:81]: I wanted to test here that there is no rubbish returned from ```bit_cast```. In case of ```float->unsigned int->float``` we won't know if wrong unsigned int was produced. I agree your option is safer regardless to device-independence, do you think we can leave only ```<type1>-><type2>-><type1>``` check and remove all ```<type1>-><type2>; if value of type2 from bit_cast == expected value of type2``` checks?

Comment (s-kanaev) [reply sycl/test/bit_cast/bit_cast.cpp:81]: > I wanted to test here that there is no rubbish returned from `bit_cast`. In case of `float->unsigned int->float` we won't know if wrong unsigned int was produced.

Well, I believe there can be some margin when converting `float -> unsigned integer`.

Comment (dm-vodopyanov) [reply sycl/test/bit_cast/bit_cast.cpp:81]: For now there is no margin for CPU and GPU both on Windows and Linux in CI but it might happen in the future for other architectures, so ok, I will introduce only ```<type1>-><type2>-><type1>``` checks.

Comment (dm-vodopyanov) [reply sycl/test/bit_cast/bit_cast.cpp]: Updated.

File: sycl/test/bit_cast/bit_cast.cpp [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+template <typename To, typename From>
+int test(const From &Value) {
+  auto ValueConvertedTwoTimes = doBitCast<From>(doBitCast<To>(Value));
+  bool isOriginalValueEqualsToConvertedTwoTimes = false;
+  if (std::is_integral<From>::value) {
+    isOriginalValueEqualsToConvertedTwoTimes = Value == ValueConvertedTwoTimes;
+  } else if ((std::is_floating_point<From>::value) || std::is_same<From, cl::sycl::half>::value) {
+    static const float Epsilon = 0.0000001f;
+    isOriginalValueEqualsToConvertedTwoTimes = fabs(Value - ValueConvertedTwoTimes) < Epsilon;
```
Comment (keryell) [inline sycl/test/bit_cast/bit_cast.cpp:43]: I am curious about how 2 bit casts could lead to non strict equality, even on floating point.

File: sycl/test/bit_cast/bit_cast.cpp [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+  ReturnCode += test<short>(cl::sycl::half(1.0f));
+
+  std::cout << "short to cl::sycl::half ...\n";
+  ReturnCode += test<cl::sycl::half>(static_cast<short>(16384));
+
+  std::cout << "int to float ...\n";
+  ReturnCode += test<float>(static_cast<int>(2));
+
+  std::cout << "float to int ...\n";
+  ReturnCode += test<int>(static_cast<float>(-2.4f));
```
Comment (keryell) [inline sycl/test/bit_cast/bit_cast.cpp:75]: Is it not already a float? If so, useless `static_cast`...

File: sycl/test/bit_cast/bit_cast.cpp [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+  std::cout << "unsigned int to float ...\n";
+  ReturnCode += test<float>(static_cast<unsigned int>(6));
+
+  std::cout << "float to unsigned int ...\n";
+  ReturnCode += test<unsigned int>(static_cast<float>(-2.4f));
```
Comment (keryell) [inline sycl/test/bit_cast/bit_cast.cpp:81]: Is it not already a float?

Otherwise when you want a precise FP value you can always use the hexadecimal floating-point literals https://en.cppreference.com/w/cpp/language/floating_literal but this seems over-kill here anyway.

File: sycl/test/bit_cast/bit_cast.cpp (hunk 11 above)
Comment (dm-vodopyanov) [reply sycl/test/bit_cast/bit_cast.cpp:43]: I made it based on @s-kanaev comment:
> check if original and result value don't differ too much
I'm not aware of such cases.
@s-kanaev do you know such cases?

File: sycl/test/bit_cast/bit_cast.cpp (hunk 12 above)
Comment (dm-vodopyanov) [reply sycl/test/bit_cast/bit_cast.cpp:75]: I made it just for readability.

File: sycl/test/bit_cast/bit_cast.cpp (hunk 13 above)
Comment (dm-vodopyanov) [reply sycl/test/bit_cast/bit_cast.cpp:81]: Just for readability. I don't see any strong point to change this but if you insist I can remove it, no problem.

File: sycl/test/bit_cast/bit_cast.cpp (hunk 11 above)
Comment (s-kanaev) [reply sycl/test/bit_cast/bit_cast.cpp:43]: > do you know such cases?

In practice I didn't met this kind of issue at the same device, but only at multiple devices. That was the point of my comment -- the same bit representation of a 4-byte `float` could result in different accuracy on multiple devices.

Maybe @keryell will correct me if I'm wrong here.

Comment (keryell) [reply sycl/test/bit_cast/bit_cast.cpp:43]: This seems to me like assignment not being perfect.

I have some colleagues doing this to spare energy on FPGA by compressing and truncating, but I guess we are not in this area...

**Comment by bader** [comment]
@dm-vodopyanov, could you update the description, please? I think it's doesn't exactly match the patch.

**Comment by dm-vodopyanov** [comment]
@bader thanks, found a mistake in description (renamed sycl::intel::bit_cast to sycl::detail::bit_cast). Everything else looks good for me. If something else should be added/removed, please tell me.
